File: gaiag/baseline/component/Reply4.java
// Gaiag --- Guile in Asd In Asd in Guile.
//
// This file is part of Gaiag.
//
// Copyright © 2014 Jan Nieuwenhuizen <janneke@gnu.org>
//
// Gaiag is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Gaiag is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Gaiag.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Reply4{
  enum Status {
    Yes, No
  };

  Boolean dummy;
  I.Status reply_I_Status;

  U.Status reply_U_Status;


  I i;
  U u;

  public Reply4() {
    dummy = false;
    i = new I();
    u = new U();
    i.getIn().done = new ValuedAction<I.Status>() {
      public I.Status action() {
        return i_done();
      }
    };
  };
  public I.Status i_done() {
    System.err.println("Reply4.i_done");
    if (true) {
      {
        U.Status s = u.getIn().what.action();
        s = u.getIn().what.action();
        if (s == U.Status.Ok) {
          Status v = fun();
          if (v == Status.Yes) reply_I_Status = I.Status.Yes;
          else reply_I_Status = I.Status.No;
        }
        else {
          Status v = this.fun_arg(Status.No);
          if (v == Status.Yes) reply_I_Status = I.Status.Yes;
          else reply_I_Status = I.Status.No;
        }
      }
    }
    return reply_I_Status;
  };
  public Status fun () {
    return Status.Yes;
  };
  public Status fun_arg (Status s) {
    return s;
  };

}

File: gaiag/baseline/Reply4.c
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2015 Paul Hoogendijk <paul.hoogendijk@verum.com>
// Copyright © 2015 Rutger van Beusekom <rutger.van.beusekom@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include "Reply4.h"

#include "locator.h"
#include "runtime.h"
#include <assert.h>

typedef enum {
	Reply4_Status_Yes, Reply4_Status_No
} Reply4_Status;




typedef struct {int size;int (*f)(Reply4*);Reply4* self;} args_i_done;




static void helper_i_done(void* args) {
	args_i_done *a = args;
	a->f(a->self);
}



static int fun(Reply4* self);
static int fun_arg(Reply4* self, int s);


static int fun(Reply4* self) {
	(void)self;
	return Reply4_Status_Yes;
}

static int fun_arg(Reply4* self, int s) {
	(void)self;
	return s;
}


static int i_done(Reply4* self) {
	(void)self;
	DZN_LOG("Reply4.i_done");
	if (true) {
		int s = self->u->in.what(self->u);
		s = self->u->in.what(self->u);
		if (s == U_Status_Ok) {
			int v = fun (self);
			if (v == Reply4_Status_Yes) self->reply_I_Status = I_Status_Yes;
			else self->reply_I_Status = I_Status_No;
		}
		else {
			int v = fun_arg(self, Reply4_Status_No);
			if (v == Reply4_Status_Yes) self->reply_I_Status = I_Status_Yes;
			else self->reply_I_Status = I_Status_No;
		}
	}
	return self->reply_I_Status;
}

static int callback_i_done(I* self) {
	args_i_done a = {sizeof(args_i_done), i_done, self->in.self};
	runtime_event(helper_i_done, &a);
	Reply4* self_ = self->in.self;
	return self_->reply_I_Status;
}


void Reply4_init (Reply4* self, locator* dezyne_locator) {
	runtime_sub_init(dezyne_locator->rt, &self->sub);
	self->dummy = false;
	self->i = &self->i_;
	self->i->in.done = callback_i_done;
	self->i->in.self = self;
	self->u = &self->u_;
	self->u->out.self = self;
}
